controller/disk_schedule_controller.go

```diff
@@ -0,0 +1,570 @@
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", obj, err))
+		return
+	}
+
+	dsc.queue.Add(key)
+}
+
+func (dsc *DiskScheduleController) enqueueReplicaDelete(obj any) {
+	replica, typeErr := eventObjToTypedObj[longhorn.Replica](obj)
+	if typeErr != nil {
+		utilruntime.HandleError(typeErr)
+		return
+	}
+
+	if replica.Spec.DiskID == "" {
+		return
+	}
+
+	diskSchedule, err := dsc.ds.GetDiskScheduleRO(replica.Spec.DiskID)
+	if err != nil {
+		if !datastore.ErrorIsNotFound(err) {
+			utilruntime.HandleError(fmt.Errorf("failed to get disk schedule %v for replica %v: %v ", replica.Spec.DiskID, replica.Name, err))
+		}
+		return
+	}
+	dsc.enqueueDiskSchedule(diskSchedule)
+}
+
+func (dsc *DiskScheduleController) enqueueBackingImageDelete(obj any) {
+	bi, typeErr := eventObjToTypedObj[longhorn.BackingImage](obj)
+	if typeErr != nil {
+		utilruntime.HandleError(typeErr)
+		return
+	}
+
+	diskUUIDs := make(map[string]struct{})
+	for diskUUID := range bi.Spec.DiskFileSpecMap {
+		diskUUIDs[diskUUID] = struct{}{}
+	}
+	for diskUUID := range bi.Status.DiskFileStatusMap {
+		diskUUIDs[diskUUID] = struct{}{}
+	}
+
+	for diskUUID := range diskUUIDs {
+		diskSchedule, err := dsc.ds.GetDiskScheduleRO(diskUUID)
+		if err != nil {
+			if !datastore.ErrorIsNotFound(err) {
+				utilruntime.HandleError(fmt.Errorf("failed to get disk schedule %v for backing image %v: %v ", diskUUID, bi.Name, err))
+			}
+			return
+		}
+		dsc.enqueueDiskSchedule(diskSchedule)
+	}
+}
+
+func (dsc *DiskScheduleController) isDiskScheduleInitialing(diskSchedule *longhorn.DiskSchedule) bool {
+	condition := types.GetCondition(diskSchedule.Status.Conditions, longhorn.DiskScheduleConditionTypeReady)
+	return condition.Status == longhorn.ConditionStatusUnknown
+}
+
+func (dsc *DiskScheduleController) syncFromAllocatedResources(diskSchedule *longhorn.DiskSchedule) error {
+	dsc.logger.Infof("Syncing disk schedule %v from allocated resources", diskSchedule.Name)
+	logger := getLoggerForDiskSchedule(dsc.logger, diskSchedule)
+
+	updateAllocation := func(specMap *map[string]int64, statusMap *map[string]*longhorn.DiskScheduledResourcesStatus, name string, size int64) {
+		if *specMap == nil {
+			*specMap = make(map[string]int64)
+		}
+		if *statusMap == nil {
+			*statusMap = make(map[string]*longhorn.DiskScheduledResourcesStatus)
+		}
+
+		(*specMap)[name] = size
+		(*statusMap)[name] = &longhorn.DiskScheduledResourcesStatus{
+			Result:        longhorn.DiskScheduledResultScheduled,
+			RequiredSize:  size,
+			ScheduledSize: size,
+		}
+	}
+
+	// sync from existing replicas
+	replicas, replicaErr := dsc.ds.ListReplicasByDiskUUID(diskSchedule.Name)
+	if replicaErr != nil {
+		return errors.Wrapf(replicaErr, "failed to list existing replicas for disk schedule %v", diskSchedule.Name)
+	}
+	for _, replica := range replicas {
+		if !isReplicaAllocatedToDisk(replica, diskSchedule) {
+			continue
+		}
```

Comment:
If a replica is unable to allocate space, we need to mark it failed. Where to fail it?